### ISSUE TITLE
Does the nfs client can do auto failover to another good nfs-ganesha server if one ganesha server is down.

### ISSUE THREAD

**Issue by lpstudy** [nfs-ganesha/nfs-ganesha#761]
We have multiple ganesha servers as our nfs proxies for one backend storage cluster. When executing mount command, the nfs client builds the connection with one ganesha server. We want to know how we can let the nfs client connects to another working nfs ganesha server automatically when current connected nfs ganesha server is down. 

Currently we are using a load-balance service for automatically failover, which exports one `vip` to the nfs client. Nfs client builds the connection with the load-balance service, will is responsible to transfer the client request to one of backend nfs ganesha servers. 
For example
``` 
                  vip                                             one of the servers, e.g. ganesha_1
nfs client  ---->     load-balancing service              ------------------->        {ganesha_1, ganehsa_2, ganesha_3} 

```
when `ganesha_1` is down, the load-balancing service can know that by `heartbeat checker` and automatically reconstruct the connection with another ganesha server, eg. `ganesha_2`.
This load-balancing service will increase the request latency because each nfs request will flow through it. We know most client in storage systems can do the automatically failover by keeping the all the alive servers in its local memory. When the client found that current proxy is down, it will switch to another good proxy server immediately. That is, does the nfs client supports this function? 

I found the nfs client can mount by a domain name and I am in doubt if we can bind  the domain name to be a list of nfs ganesha servers. So that when the client found current nfs ganesha server is down, it can switch to another good ganesha server. I have googled the information but can not find something usefull. Need your guys' help.

**Comment by dang** [comment]
NFS doesn't have auto-failover on the client side in the protocol.  Instead, the server fails over, and the clients re-connect to the new server.  It was designed originally (40+ years ago) with the idea that server failure would be rare, and could have a relatively high latency recovery operation.  If you want load balancing, then you cannot escape the setup you have.  However, if all you want is HA, you can use a setup like pacemaker/corosync or ctdb, and assign VIPs directly to the NFS servers.  CTDB would then fail the VIP over to another running server when a server fails.  This won't remove the latency during failover, but will remove the latency introduced by the load balancer into normal ops.

**Comment by ffilz** [comment]
All the NFS active/active HA solutions I know details of (including uses of Ganesha's GPFS and Gluster FSALs) use VIP failover as Daniel mentioned. A load balancer in front of an NFS server has problems if a given client can be re-allocated to a different server without a failure event, in which case neither the client nor the server understand that the client has moved, and any state (open files, locks, delegations, pnfs layouts) is left on the old server (to eventually be timed out in NFSv4 since the client will no longer be sending lease renewal). Actually, with NFSv4, the client and server will know there was some issue because the server won't recognize the clientid the client uses and will return a bad clientid error which will cause the client to dump the state.

The VIP system eliminates latency on a per request basis, only causing latency due to failover recovery (which is unavoidable without an active/active system that shares ALL the state among all the nodes).

**Comment by lpstudy** [comment]
> NFS doesn't have auto-failover on the client side in the protocol. Instead, the server fails over, and the clients re-connect to the new server. It was designed originally (40+ years ago) with the idea that server failure would be rare, and could have a relatively high latency recovery operation. If you want load balancing, then you cannot escape the setup you have. However, if all you want is HA, you can use a setup like pacemaker/corosync or ctdb, and assign VIPs directly to the NFS servers. CTDB would then fail the VIP over to another running server when a server fails. This won't remove the latency during failover, but will remove the latency introduced by the load balancer into normal ops.

Hi, thanks for your reply. 
I am wondering if we can let the nfs client automatically switch to another ganesha server if current server failed.
We use the mount commands like this
```mount -t nfs4 -o vers=1 ip1;ip2;ip3:/ /mnt/nas```
Here the ip1;ip2;ip3 are the ganesha address list.

Question 1:
We wish we have the setup of ```client -> {ip1, ip2, ip3}```
The nfs client will randomly select one ganesha server (eg. ip1) and connect to it. If ip1 is corrupted, nfs client will lose the connection with ip1 and automatically mount to another ganesha server (eg. ip2).  Whether we can achieve this by modifying the nfs client in kernel code. If it can be done, could you give me some advice? we are woking on the kernel codes of nfs client.

Question 2: 
with the current working setup of ```client --(vip)---> load balancer -> {ip1, ip2, ip3}```
In our case, the load balancer will reply the nfs client a reset connection response and redirect the request to ip2.
The state for client in ip1 will be eventually timeouted as you talked about. I am wonderring if ip2 can recover the stats of that client in ip2. We capture the nfs network packet and found that the client will exchange the client id and create a new session in ip2 because ip2 can not find that client id.  I do not know if the nfs client will recover the opened file handles and locks that have been done in ip1.
@ffilz

**Comment by dang** [comment]
I do not believe this is possible with NFS.  The issue is that some state is stored on the server, and some state is stored on the client, and those states have to be in sync.  This means that, if the client has a file locked on a server, for example, and that server goes down, it needs to restore that state with the new server.  But this requires all the servers in the cluster to enter Grace, and all the clients of the old server to resync with the new server, before Grace can be lifted, or the lock state will be lost.  This can cause data corruption in the file.

Basically, the NFS protocol is not designed to allow a client to transparently move between servers.

**Comment by ffilz** [comment]
For question1, the closest is that NFSv4 does have an fslocations attribute that allows the server to tell the client other servers that host the same file systems. I know there was some looking at this for fail over but actually migrating state (opens and locks etc.) is tricky and I'm not sure anyone ever came up with a workable solution.

For question2, one issue is that the client will close idle connections, though it would seem that an idle connection would not have NFSv4.x state (otherwise it would need to be sending lease renewal which would seem to preclude declaring the connection idle). If something forced the loss of connection, the client would definitely try to reclaim the state, however, the server as currently designed would not be prepared to allow reclaiming that state, and if the ip1 server was actually still up, the original state would still be present and cause conflicts. Back in 2000/2001 I did work on a solution (NFSv3 only) for migrating state from one server node to another, but it still was due to a VIP moving from one host to another and was definitely NFSv3 only. FSAL_CEPH does have capability to have a new Ganesha instance come up prior to CephFS discarding the state from the previous instance but that is currently not used with multiple VIPs. Ganesha in that case has only a single IP address and a new instance is started if the previous instance fails and then that new instance protects the state while the clients reclaim it, so not really migrating state, just protecting it so there isn't a window where a failure presents an opportunity for a different client to get conflicting state.

**Comment by lpstudy** [comment]
> I do not believe this is possible with NFS. The issue is that some state is stored on the server, and some state is stored on the client, and those states have to be in sync. This means that, if the client has a file locked on a server, for example, and that server goes down, it needs to restore that state with the new server. But this requires all the servers in the cluster to enter Grace, and all the clients of the old server to resync with the new server, before Grace can be lifted, or the lock state will be lost. This can cause data corruption in the file.
> 
> Basically, the NFS protocol is not designed to allow a client to transparently move between servers.

I am wonderring how the state can be recovered by using vip mode.  That is in the case that ```client --(vip)---> load balancer -> {ip1, ip2, ip3}.```
First the client is connected to ip1 and both the client and ip1 have the state (eg. lock state). And then at some point, ip1 is corrupted, and the client is reconnected to ip2. How the client can recover the lost lock state in ip1?  If the client can recover the lost state in ip1 in that case, why the client can not fail over to ip2 and recover the lost state that have been stored in ip1?
I still can not get the key point.

**Comment by lpstudy** [comment]
> For question1, the closest is that NFSv4 does have an fslocations attribute that allows the server to tell the client other servers that host the same file systems. I know there was some looking at this for fail over but actually migrating state (opens and locks etc.) is tricky and I'm not sure anyone ever came up with a workable solution.
> 
> For question2, one issue is that the client will close idle connections, though it would seem that an idle connection would not have NFSv4.x state (otherwise it would need to be sending lease renewal which would seem to preclude declaring the connection idle). If something forced the loss of connection, the client would definitely try to reclaim the state, however, the server as currently designed would not be prepared to allow reclaiming that state, and if the ip1 server was actually still up, the original state would still be present and cause conflicts. Back in 2000/2001 I did work on a solution (NFSv3 only) for migrating state from one server node to another, but it still was due to a VIP moving from one host to another and was definitely NFSv3 only. FSAL_CEPH does have capability to have a new Ganesha instance come up prior to CephFS discarding the state from the previous instance but that is currently not used with multiple VIPs. Ganesha in that case has only a single IP address and a new instance is started if the previous instance fails and then that new instance protects the state while the clients reclaim it, so not really migrating state, just protecting it so there isn't a window where a failure presents an opportunity for a different client to get conflicting state.

Do we need to persist the nfs state in ganesha? I always think that we need to persist the nfs state, but now I am not sure. I have done some experiments with the vip setup. 
``` client -----> (vip) load balancer -> {ip1, ip2}.```

nfs client is first connected to `ip1`, and I opened a file `f1` and kept appending data into `f1`. At some point, I stop the ganesha ip1 server and load balancer knows that `ip1` is down and will redirect the network flow to `ip2`. 

The writting request for file `f1` will reach to `ip2` as detailed in the followings. The ip2 return BadSession and stale client id to the client because this client is not in `ip2`. Then the client starts to construct a new client id  and a new session in `ip2`. Also the client reopened the file `f1`. By this way, the client recovered the file state for `f1` from `ip1` to `ip2`.

So I think I do not need the persist the nfs state in the ganesha server. But I heard someone that We need to persist it. Could you give me an explanation? Or some states needed to be persisted, some didn't?

```
30765 2021-11-18 15:05:27.146827366 10.145.105.211 → 10.8.6.98    NFS 594 V4 Call **WRITE** StateID: 0x7396 Offset: 0 Len: 289
30767 2021-11-18 15:05:27.147294933    10.8.6.98 → 10.145.105.211 NFS 114 V4 Reply (Call In 30765) SEQUENCE Status: NFS4ERR_BADSESSION
30769 2021-11-18 15:05:27.147399478 10.145.105.211 → 10.8.6.98    NFS 174 V4 Call DESTROY_SESSION
30770 2021-11-18 15:05:27.147612338    10.8.6.98 → 10.145.105.211 NFS 114 V4 Reply (Call In 30769) DESTROY_SESSION Status: NFS4ERR_BADSESSION
30788 2021-11-18 15:05:27.150474915 10.145.105.211 → 10.8.6.98    NFS 274 V4 Call CREATE_SESSION
30792 2021-11-18 15:05:27.150652649    10.8.6.98 → 10.145.105.211 NFS 114 V4 Reply (Call In 30788) CREATE_SESSION Status: NFS4ERR_STALE_CLIENTID
30863 2021-11-18 15:05:27.153267887 10.145.105.211 → 10.8.6.98    NFS 310 V4 Call **EXCHANGE_ID**
30865 2021-11-18 15:05:27.153429879    10.8.6.98 → 10.145.105.211 NFS 194 V4 Reply (Call In 30863) EXCHANGE_ID
30914 2021-11-18 15:05:27.156877252 10.145.105.211 → 10.8.6.98    NFS 274 V4 Call **CREATE_SESSION**
30921 2021-11-18 15:05:27.157124931    10.8.6.98 → 10.145.105.211 NFS 194 V4 Reply (Call In 30914) CREATE_SESSION
31016 2021-11-18 15:05:27.164602103 10.145.105.211 → 10.8.6.98    NFS 206 V4 Call **PUTROOTFH** | GETATTR
31046 2021-11-18 15:05:27.178103485    10.8.6.98 → 10.145.105.211 NFS 182 V4 Reply (Call In 31016) PUTROOTFH | GETATTR
31048 2021-11-18 15:05:27.178171317 10.145.105.211 → 10.8.6.98    NFS 198 V4 Call RECLAIM_COMPLETE
31049 2021-11-18 15:05:27.178322623    10.8.6.98 → 10.145.105.211 NFS 158 V4 Reply (Call In 31048) RECLAIM_COMPLETE
31050 2021-11-18 15:05:27.178356662 10.145.105.211 → 10.8.6.98    NFS 218 V4 Call TEST_STATEID
31051 2021-11-18 15:05:27.178568285    10.8.6.98 → 10.145.105.211 NFS 166 V4 Reply (Call In 31050) TEST_STATEID
31115 2021-11-18 15:05:27.186396341 10.145.105.211 → 10.8.6.98    NFS 334 V4 Call **OPEN** DH: 0x1eab11ed/
31134 2021-11-18 15:05:27.188019674    10.8.6.98 → 10.145.105.211 NFS 382 V4 Reply (Call In 31115) OPEN StateID: 0xb24a
35523 2021-11-18 15:05:32.250567485 10.145.105.211 → 10.8.6.98    NFS 218 V4 Call TEST_STATEID
35524 2021-11-18 15:05:32.250923250    10.8.6.98 → 10.145.105.211 NFS 166 V4 Reply (Call In 35523) TEST_STATEID
35526 2021-11-18 15:05:32.251007648 10.145.105.211 → 10.8.6.98    NFS 594 V4 Call **WRITE** StateID: 0xb7e7 Offset: 0 Len: 289
35532 2021-11-18 15:05:32.252979002    10.8.6.98 → 10.145.105.211 NFS 254 V4 Reply (Call In 35526) WRITE
```

**Comment by dang** [comment]
That's not a clean transition, it's a failure.  The server failed, the client tore down it's connection, and created a new one.  All outstanding changes were (potentially) lost, corrupting data in flight.  In order for it to be a clean transition, the servers on the cluster need to be put into grace when the VIP fails over.

**Comment by ffilz** [comment]
OK, some thoughts...

So NFS state consists of:

byte range locks (NFS v3 and NFSv4.0+)

opens "share reservations", these may have access mode and deny mode (NFSv4.0+ only for this context - NFSv3 does have share reservations by way of NLM however they are not recoverable state)

delegations (NFSv4.0+ only)

layouts (NFSv4.1+)

Any state beyond an open with no deny mode establishes a contract with the server for some specific behavior.

If the client application is using byte range locks or deny modes to prevent another client from accessing the file data in an inconsistent way, then "loss of state" means this contract is broken.

Delegations are a contract with the client that allows the client to cache data (read cache or write back cache) and even manage byte range locks and deny modes. Again, if state is lost, then the client can no longer rely on it's cache.

Layouts are a contract with the server that allows the client to perform I/O independently of the connection with the specific server. This contract again requires that the server manage what other access is going on to protect this I/O from corruption so loss of state again means the contract is broken.

Now if you application is not relying on any of this "state" to assure correctness of data during possible concurrent access from another client, loss of state is not so much a problem. Now there is still the problem as to whether a write completes before the node crash, however the client knows which writes may be incomplete (has not got a write response, or has not got a commit response after a series of unstable writes). In such case, the client knows it can not trust the data was written properly to stable storage, and can re-write the data to the new node after fail over. There is some complexity and there are some concerns with ordering of I/O, however, in the absence of state, there was never any guarantee of ordering of I/O.

Persisting the state we are talking about here is an interesting proposition, but in order for it to be remotely performant there must be a very fast shared storage to persist the state to.

**Comment by dang** [comment]
The problem with this is that it assumes you wrote the client.  If you're using a pre-existing client (such as the Linux kernel client), then it does what it does, and likely takes full advantage of these abilities when it can.  So you have to assume that at least some of these features are in use.

**Comment by ffilz** [comment]
Well, a Linux client won't ask for deny modes, and it will only use byte range locks when the application uses them (and you can assure it doesn't by mounting -o nolock), leaving delegations really the only issue. And if you're not using deny modes or byte range locks, you shouldn't be sharing files between clients in a way that delegations could introduce any more opportunities for data corruption. And there's something to be said for if you truly don't need any state to just use NFSv3...

**Comment by ffilz** [comment]
Can we close this or do you still have open questions?

**Comment by ffilz** [comment]
Closing old issues that are not planned for any resolution.

Please re-open with new information if this issue is still of concern.
